### DISCUSSION TITLE
Use alpine in a variable or a function ¿Some difference?

### DISCUSSION THREAD

**Discussion by xcesaralejandro** [alpinejs/alpine#1099] [4. General]
I am just getting into alpine and I have noticed that all the examples have a function for x-data that returns an object, while if we see other frameworks like vuejs, most of them are instantiated in a variable, being that vue is composed of an object.

Is there any underlying reason to declare x-data properties as a return object in a function instead of a variable? or both forms are equally valid.

````
var textarea_format = {
            resize: function(){
                this.$el.style.height = '5px';
                this.$el.style.height = (this.$el.scrollHeight + 5) + 'px'
            }
        }

    function textarea_format(){
        return {
            resize: function(){
                this.$el.style.height = '5px';
                this.$el.style.height = (this.$el.scrollHeight + 5) + 'px'
            }
        }
    }
````

**Comment by KevinBatdorf** [comment]
If you have a choice, use a function. Alpine V2 will evaluate it slightly faster:

See: https://github.com/alpinejs/alpine/blob/95cff65be3a91976222f7a72a4d3ab9f52b281bc/src/utils.js#L88

**Comment by SimoTod** [comment]
Functions make sure you get a fresh copy of the object for every components when you use it. A  variable, if containing properties, would be partially shared (but without the reactive part so it will lead to bugs). See https://codepen.io/SimoTod/pen/rNWmjZR
If you click the first +, counter becomes 1 for both components but it only gets visually updated on the first one. If you click the second +, counter becomes 2 but it only gets visually updated on the second component. The final user would think that counter is still 0 for the second component and it would expect it to increment to 1

**Reply by KevinBatdorf** [reply]
That's because you're passing the reference to the object, right?

**Reply by SimoTod** [reply]
Yeah, but each component creates a different proxy around it so they only react to their own changes
